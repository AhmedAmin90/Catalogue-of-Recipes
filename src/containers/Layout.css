.Layout {
    width: 90%;
    display: flex;
    justify-content: space-between;
    flex-wrap: wrap;
    margin: 0 auto;
}