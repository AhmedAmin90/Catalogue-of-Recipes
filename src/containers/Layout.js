/* eslint-disable consistent-return , no-restricted-syntax */
import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import Card from './Card';
import CategoryFilter from '../components/CategoryFilter';
import * as actions from '../actions/index';
import './Layout.css';

const Layout = ({ selectedCategory, foods, filteredFood }) => {
  const dispatch = useDispatch();

  const renderSelectedCategory = useSelector((state) => state.selectedReducer);
  const renderFoods = useSelector((state) => state.foodReducer);
  const renderFilteredFood = useSelector((state) => state.filterReducer);

  const handleFilterChange = (category) => {
    dispatch(actions.CHANGE_FILTER(category));
  };

  const selectedFoods = (category) => {
    if (category === 'All') {
      if (!foods) {
        return renderFoods.map((oneRecipe) => (
          <Link key={oneRecipe.id} to={`/recipes/${oneRecipe.id}`}>
            <Card
              key={oneRecipe.id}
              id={oneRecipe.id}
              name={oneRecipe.name}
              img={oneRecipe.image}
            />
          </Link>
        ));
      }
      return foods.map((oneRecipe) => (
        <Link key={oneRecipe.id} to={`/recipes/${oneRecipe.id}`}>
          <Card key={oneRecipe.id} id={oneRecipe.id} name={oneRecipe.name} img={oneRecipe.image} />
        </Link>
      ));
    }
    if (!selectedCategory) {
      for (const cat of renderSelectedCategory) {
        if (cat[category]) {
          return cat[category].map((food) => (
            <Link key={food.idMeal} to={`/recipes/${food.idMeal}`}>
              <Card
                key={food.idMeal}
                id={food.idMeal}
                name={food.strMeal}
                img={food.strMealThumb}
              />
            </Link>
          ));
        }
      }
    } else {
      for (const cat of selectedCategory) {
        if (cat[category]) {
          return cat[category].map((food) => (
            <Link key={food.idMeal} to={`/recipes/${food.idMeal}`}>
              <Card
                key={food.idMeal}
                id={food.idMeal}
                name={food.strMeal}
                img={food.strMealThumb}
              />
            </Link>
          ));
        }
      }
    }
  };

  const handleAllCategroy = () => {
    handleFilterChange('All');
  };
  const loading = () => {
    if (!selectedCategory || !foods || !filteredFood) {
      return renderSelectedCategory.length === 0
        ? (

          <div className="Loading">
            <h1>Please Wait; Our delicious foods are Loading ... </h1>
            <p>
              This built with
              <i className="fas fa-heart" />

              by Ahmed Amin

            </p>
            <ul className="social-icons">
              <li>

                <a href="https://twitter.com/AhmedAmin12383" target="_blank" rel="noreferrer">
                  <i className="fab fa-twitter" />
                </a>

              </li>
              <li>

                <a href="https://www.linkedin.com/in/web-developer/" target="_blank" rel="noreferrer">

                  <i className="fab fa-linkedin-in" />

                </a>

              </li>
              <li>

                <a href="https://github.com/AhmedAmin90" target="_blank" rel="noreferrer">

                  <i className="fab fa-github" />

                </a>
              </li>
              <li>

                <a href="https://angel.co/u/ahmed-amin-22" target="_blank" rel="noreferrer">

                  <i className="fab fa-angellist" />

                </a>
              </li>
              <li>

                <a href="https://www.facebook.com/ahmed.amin.7564" target="_blank" rel="noreferrer">

                  <i className="fab fa-facebook-f" />

                </a>
              </li>
            </ul>
          </div>
        )

        : (
          <div>
            <nav>
              <CategoryFilter handleFilter={handleFilterChange} value={renderFilteredFood} />
              <button type="button" className="Layout-all-categories" onClick={handleAllCategroy}>All categories</button>
            </nav>
            <div className="Layout">
              {selectedFoods(renderFilteredFood)}
            </div>
          </div>
        );
    }
    return (
      <div>
        <nav>
          <CategoryFilter handleFilter={handleFilterChange} value={filteredFood} />
          <button type="button" className="Layout-all-categories" onClick={handleAllCategroy}>All categories</button>
        </nav>
        <div className="Layout">
          {selectedFoods(filteredFood)}
        </div>
      </div>
    );
  };
  return (
    <div>{loading()}</div>

  );
};

Layout.defaultProps = {
  selectedCategory: null,
  filteredFood: null,
  foods: null,
};

Layout.propTypes = {
  selectedCategory: PropTypes.instanceOf(Array),
  filteredFood: PropTypes.string,
  foods: PropTypes.instanceOf(Array),
};

export default Layout;
