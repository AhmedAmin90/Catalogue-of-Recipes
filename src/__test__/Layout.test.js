import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import { create } from 'react-test-renderer';
import { Provider } from 'react-redux';
import {
  BrowserRouter as Router,
} from 'react-router-dom';
import store from '../index';
import Layout from '../containers/Layout';

test('renders correctly', () => {
  const tree = create(<Provider store={store}><Router><Layout /></Router></Provider>).toJSON();
  expect(tree).toMatchSnapshot();
});

describe('rendered Layout', () => {
  let renderedComponent;
  beforeEach(() => {
    const selectedCat = [{
      Beef: [{
        strMeal: 'Beef and Mustard Pie',
        strMealThumb: 'https://www.themealdb.com/images/media/meals/sytuqu1511553755.jpg',
        idMeal: '52874',
      }],
    },
    ];

    const foods = [{
      id: '52874',
      name: 'Beef and Mustard Pie',
      image: 'https://www.themealdb.com/images/media/meals/sytuqu1511553755.jpg',
    }];

    const filteredFood = 'Beef';

    renderedComponent = render(
      <Provider store={store}>
        <Router>
          <Layout foods={foods} selectedCategory={selectedCat} filteredFood={filteredFood} />
        </Router>
      </Provider>,
    );
  });
  it('has a div with class Layout', () => {
    const { container } = renderedComponent;
    const appElement = container.querySelector('.Layout');
    expect(appElement).toBeInTheDocument();
  });

  it('has a card with card img background', () => {
    const { container } = renderedComponent;
    const appElement = container.querySelector('.Card');
    expect(appElement).toBeInTheDocument();
    expect(appElement).toHaveStyle('background-image: url(\'https://www.themealdb.com/images/media/meals/sytuqu1511553755.jpg\')');
  });

  it('has a card with card img background', () => {
    const { container } = renderedComponent;
    const appElement = container.querySelector('.Card-name');
    expect(appElement).not.toBeFalsy();
    expect(appElement.innerHTML).toBe('Beef and Mustard Pie');
  });

  it('without a div with class loading', () => {
    const { container } = renderedComponent;
    const appElement = container.querySelector('.Loading');
    expect(appElement).toBeNull();
    // expect(getByText("Click Me").href).toBe("/recipes/52874")
  });

  it('test link href ', () => {
    const { container } = renderedComponent;
    const appElement = container.querySelector('.Layout').querySelector('a');
    expect(appElement.href).toBe('http://localhost/recipes/52874');
  });

  it('Presence of the link ', () => {
    const { container } = renderedComponent;
    const appElement = container.querySelector('.Layout').querySelector('a');
    expect(appElement).not.toBeFalsy();
  });

  it('has a has a name of food with : ', () => {
    const { getByText } = renderedComponent;
    const appElement = getByText('Beef and Mustard Pie');
    expect(appElement).toBeInTheDocument();
  });

  it('has a has a name of food with : ', () => {
    const { getByText } = renderedComponent;
    const appElement = getByText('Beef and Mustard Pie');
    expect(appElement).not.toBeNull();
  });

  it('has a a name of category with Beef', () => {
    const { getByText } = renderedComponent;
    const appElement = getByText('Beef');
    expect(appElement).toBeInTheDocument();
  });
});

describe('rendered Main', () => {
  let renderedComponent;
  beforeEach(() => {
    renderedComponent = render(
      <Provider store={store}>
        <Router>
          <Layout />
        </Router>
      </Provider>,
    );
  });

  it('has a div with class loading', () => {
    const { container } = renderedComponent;
    const appElement = container.querySelector('.Loading');
    expect(appElement).toBeInTheDocument();
  });

  it('without a div with class Layout', () => {
    const { container } = renderedComponent;
    const appElement = container.querySelector('.Layout');
    expect(appElement).toBeNull();
  });

  it('has a div a waiting text ', () => {
    const { getByText } = renderedComponent;
    const appElement = getByText('Please Wait; Our delicious foods are Loading ...');
    expect(appElement).toBeInTheDocument();
  });

  it('No name food in loading pag', () => {
    const { container } = renderedComponent;
    const appElement = container.querySelector('.Layout-all-categories');
    expect(appElement).not.toBeInTheDocument();
  });
});
