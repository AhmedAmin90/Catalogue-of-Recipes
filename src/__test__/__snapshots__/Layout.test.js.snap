// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="Loading"
  >
    <h1>
      Please Wait; Our delicious foods are Loading ... 
    </h1>
    <p>
      This built with
      <i
        className="fas fa-heart"
      />
      by Ahmed Amin
    </p>
    <ul
      className="social-icons"
    >
      <li>
        <a
          href="https://twitter.com/AhmedAmin12383"
          rel="noreferrer"
          target="_blank"
        >
          <i
            className="fab fa-twitter"
          />
        </a>
      </li>
      <li>
        <a
          href="https://www.linkedin.com/in/web-developer/"
          rel="noreferrer"
          target="_blank"
        >
          <i
            className="fab fa-linkedin-in"
          />
        </a>
      </li>
      <li>
        <a
          href="https://github.com/AhmedAmin90"
          rel="noreferrer"
          target="_blank"
        >
          <i
            className="fab fa-github"
          />
        </a>
      </li>
      <li>
        <a
          href="https://angel.co/u/ahmed-amin-22"
          rel="noreferrer"
          target="_blank"
        >
          <i
            className="fab fa-angellist"
          />
        </a>
      </li>
      <li>
        <a
          href="https://www.facebook.com/ahmed.amin.7564"
          rel="noreferrer"
          target="_blank"
        >
          <i
            className="fab fa-facebook-f"
          />
        </a>
      </li>
    </ul>
  </div>
</div>
`;
